docs/source/blueprints/BP002.rst

```diff
@@ -0,0 +1,220 @@
+------------------------------------------------
+
+Another way to create the requirements representation as JSON files is by
+writing a Python executable. This approach makes the requirements
+representation flexible, by allowing the use of Python variables and code that
+may change the parameters values for the requirements, depending on the
+environment the Python code runs.
+
+The following example shows a requirement that depends on the architecture
+the test is running::
+
+        #!/usr/bin/python3
+
+        import os
+        import json
+
+        requirements = [
+            {"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"},
+            {"type": "vmimage", "distro": "fedora", "version": 31, "arch": os.uname()[4]},
+            {"type": "package", "package": "lvm"}
+        ]
+
+        print(json.dumps(requirements))
+
+Requirements representation as metadata on test docstring
+---------------------------------------------------------
+
+Test writers may want to add the requirements of a test into de test code. The
+option proposed here allows the use of metadata on test docstrings to represent
+the requirements list.
+
+Below is an example of how to define requirements as metadata on docstrings::
+
+        def test_something(self):
+        '''
+            :avocado requirement: [{"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}]
+            :avocado requirement: [{"type": "vmimage", "distro": "fedora", "version": 31, "arch": "x86_64"}]
+            :avocado requirement: [{"type": "package", "package": "lvm"}]
+        '''
+           <test code>
+
+Requirements files location
+===========================
+
+It may be useful for test writers to define a standard source location for the
+requirements JSON files and the requirements Python executable.
+
+This blueprint defines the default location for those files into a requirements
+directory preceded by the test file name. It is also supported the use of
+sub-directories using the name of a specific test to define requirements for
+that test.
+
+The following file tree is an example of possible use for requirements
+directories for a test::
+
+        cabort.py
+        cabort.py.requirements/
+        ├── CAbort.test_2
+        │   └── requirements.py
+        └── requirements.json
+
+In this case, all the tests on `cabort.py`, except for `CAbort.test_2`, use the
+`requirements.json` file located at `cabort.py.requirements`. The `CAbort.test_2`
+test uses its own `requirements.py` located at `CAbort.test_2` directory inside
+the requirements root directory.
+
+Requirements files command-line parameter
+=========================================
+
+It is also possible to use a command-line parameter to define the location of
+the requirements file. For that, this blueprint defines the parameter
+`--requirements-file` followed by the location of the requirements file. As a
```

Comment:
Are you proposing an "execution wide" requirements file?  Or are you proposing the configuration of the pattern/location of the JSON file that will be used?

Reply:
I'm proposing a command-line option that supersedes all the other requirements definitions possibilities, just like the `--config` parameter, with an explicit requirements file. The use case I'm thinking here is a test that already has a requirements config file, but is failing for an unknown reason. Instead of changing the original requirements config file, the person debugging it may want to use a temporary requirements config file as a command-line option to test alternative URLs for a file, a different package version, or whatever is needed.

Command-line example:

`avocado run --requirements-file /tmp/requirements.json passtest.py`

Anyway, from my point of view, this feature is a "plus." It should have the attention after all the other requirements resolver features are available, but I think it is essential to have its definition in the blueprint.

Reply:
Yes, I think this is an extra feature that can be added on top of the base functionality we're focusing on.  My impression is that the most common use cases are more important *by far*, and thinking about these can steal your focus on those.

Also my belief is that for the less commonly used features, we should soon tell users to hack their own jobs, and maybe this will be as simple as activating a different "requirements resolver" or with different options.  What I mean is something similar to `examples/job/podman.py`, but imagine we have a `examples/job/requirements.py`:

```
#!/usr/bin/env python3

import sys
from avocado.core.job import Job

config = {'references': ['/bin/true'],
          'requirements_resolver': None,
          'requirements': 'YOUR_REQUIREMENTS_HERE'}

with Job(config) as j:
    sys.exit(j.run())
```




Only after the use case is deemed common, we should incorporate it in a fully supported feature by means of a command line option.